### ISSUE TITLE
mkfs.xfs crashes dmdedup

### ISSUE THREAD

**Issue by venkrishraja** [dmdedup/dmdedup#5]
As suggested by the various OOPS dumps taken, there seems to be a race bug that seems to crash all over the code.

There was one particular oops dump,

[83874.337417] kernel BUG at block/bio.c:528!

Which comes from the BIO_BUG_ON in block/bio.c:bio_put():

I tried printing the ref count of bio and clone at different places before the crash. All seem to return the value 1, in all the places. I also tried putting bio_get() in create_bio() and bio_put() in my_endio() but that didnt seem to have any effect on the BUG.

**Comment by ezk** [comment]
Can show you us the sample patch you've used with get/put_bio so we can verify it?

With that patch in place, can you again run the system 10+ times and analyze whether the OOPS shows a similar or different pattern?  It's possible we fixed part of the race, but not in all places, so we want to check if the oops patterns change.

BTW, I think we should also submit a separate patch that uses the standard zeroing of a bio function, rather than our own.  Then repeat the tests.

**Comment by ezk** [comment]
Can you also upload a file with all the oops traces you've seen, each time you fix or change the code?  I'd like to have the history of these bugs in a convenient place to review.

**Comment by venkrishraja** [comment]
Sure, I'll update them tonight.

**Comment by sectorsize512** [comment]
Hey, guys, since you seem to start looking into this issue more intensively I wanted to let you know what I dig out so far:

1) I was able to narrow down the bug reproducibility to this small program:

```
#define _GNU_SOURCE

#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define BUFSIZE  1024 * 1024

int main()
{
        char *buf = (char *)aligned_alloc(512, BUFSIZE);
        int ret;
        int fd;

        memset(buf, 0, BUFSIZE);

        fd = open("mydedup", O_RDWR|O_EXCL|O_DIRECT);
        if (fd < 0) {
                printf("Fd: %d\n", fd);
                return fd;
        }

        ret = pwrite(fd, buf, 8192, 0);

        return ret;
}
```

Sometimes you need to run it several times to reproduce the bug.

2) The problem is most probably with read-on-write path which is implemented in -rw.c file.

3) At least one problem is not a race but memory corruption. Specifically, merge_data() function writes to much and to a wrong location when merging it. The reason is because it is not using new bio_iter field that was introduced in 3.13. Here is the _dirty_ patch that fixes the issue for the small program above:

```
--- a/dm-dedup-rw.c
+++ b/dm-dedup-rw.c
@@ -55,6 +55,7 @@ static int merge_data(struct dedup_config *dc, struct page *page,
        sector_t bi_sector = bio->bi_iter.bi_sector;
        void *src_page_vaddr, *dest_page_vaddr;
        int position, err = 0;
+       int len;

        /* Relative offset in terms of sector size */
        position = sector_div(bi_sector, dc->sectors_per_block);
@@ -67,12 +68,17 @@ static int merge_data(struct dedup_config *dc, struct page *page,
        src_page_vaddr = page_address(bio->bi_io_vec->bv_page);
        dest_page_vaddr = page_address(page);

+       // XXX: bv_offset might not be updated
        src_page_vaddr = src_page_vaddr + bio->bi_io_vec->bv_offset;
        /* Locating the right sector to merge */
        dest_page_vaddr = dest_page_vaddr + (to_bytes(position));

        /* Merging Data */
-       memmove(dest_page_vaddr, src_page_vaddr, bio->bi_io_vec->bv_len);
+       len = bio->bi_iter.bi_size;
+       printk(KERN_ERR "memmove len: %d position: %lu\n", len, to_bytes(position));
+       memmove(dest_page_vaddr, src_page_vaddr, len);
 out:
        return err;
 }
```

4) This does not fix the whole mkfs.xfs run though (but it runs much longer now without crashes with this fix). But I'm not surprised - my patch is dirty, it does not cover all the cases, and moreover there are other places where bi_iter is not used and bio vector is accessed directly. In short, IMHO, all -rw.c need to have a major revision to use bi_iter properly. (other code need to be checked as well).

I won't be able to look at it further for about a week or so.

**Comment by venkrishraja** [comment]
I have created a new pull request to show the patch that I am testing with.
PFA the oops dumps taken after the patch.

Now the crash seems to be happening at merge_data() consistently.
[oopsdump.txt](https://github.com/dmdedup/dmdedup/files/223140/oopsdump.txt)

**Comment by sectorsize512** [comment]
Yes, there is at least one issue in mege_data() - len can be too larged and we corrupt memory. That is fixed in the patch above (the fix is very dirty), but mkfs.xfs still fails - maybe in different place now.

**Comment by venkrishraja** [comment]
I have been reading about struct bio. 

In each given block I/O operation, there are bi_vcnt vectors in the bio_vec array starting with bi_io_vec. As the block I/O operation is carried out, the bi_idx field is used to point to the current index into the array. 

bi_idx resides in the structure bi_iter for each bio.

The bio_vec array has bv_page which is the struct page object, bv_offset - the byte offset within the page where the buffer resides and the buffer length bv_len.

In rw.c, we are not using the iterator bi_idx anywhere and we are directly referring to the bv_page like this:

bio->bi_io_vec->bv_page;
bio->bi_io_vec->bv_offset;
bio->bi_io_vec->bv_len;

in merge data() and other functions when it should actually be like this?

int index = bio->bi_iter.bi_idx;
bio-> bi_io_vec[index]->bv_page;

Also, since there is one struct page for each bio_vec array, the maximum data that can be processed at a time is 4k? I am not sure about this since there is not much help in this regard online.

**Comment by sectorsize512** [comment]
1) Yes, at the time when the code was initially written bi_iter did not exist. So, the code need to be updated to use the iterator.

2) Take a look at various macroses in linux/bio.h that allow to access bio's fields without going into bi_iter directly. (e.g., bio_iter_page).

3) In device-mapper framework we set the unit of mapping to 4KB. This should guarantee that every bio is not larger than 4KB (but asserts on that are still a good idea). Now, are 4KB split across several vectors or not - might not be guaranteed.

HTH,
Vasily

**Comment by ezk** [comment]
OK thanks.  At least we can prevent bios greater than 4kb.

Venkat, have you tried to make changes to the code and see what works?  Do you have proposed patches for us to try?

Erez.

**Comment by venkrishraja** [comment]
No I haven't made any significant changes to the code. I wanted to clarify a few things before that.
There is a macro 'bio_for_each_segment' to iterate through the bi_io_vec arrays.
1. While creating a new clone bio, we are creating just one page and adding it to the bio structure while we should be replicating the bi_io_vec array which is an array of pages.
2. The only places where we have accessed bi_io_vec directly are merge_data and copy_pages. These functions take one new page and bio as input and copy the first page from bio to the new page. This function has to be called several times from inside the loop to copy each bio_vec array.

Or would it be wiser to call the function once and iterate from inside the function?
Thoughts?

**Comment by ezk** [comment]
Inline.

> On Apr 24, 2016, at 4:13 AM, Venkatakrishnan notifications@github.com wrote:
> 
> No I haven't made any significant changes to the code. I wanted to clarify a few things before that.
> There is a macro 'bio_for_each_segment' to iterate through the bi_io_vec arrays.
> 
> I’m not sure if below you are stating what we are doing, proposing a new way, or just asking Qs.
> 
>   • While creating a new clone bio, we are creating just one page and adding it to the bio structure while we should be replicating the bi_io_vec array which is an array of pages.

When we create a new bio, it has to be properly formatted.  If every bio requires a vector array, then we should do so too.  Recall we only handle 4KB units, so at most we’ll need only one page. Still, we may need to create a vector of just one element and stuff that page into it.  And we’ll need to ensure that all the other params are correct: index into the array, etc.

I don’t think we need to replicate exactly the vector we get from above, as that could be split into several small chunks.

On our meeting on Monday, I’d like you to draw for me a picture on the whiteboard showing the relationship of these struct bio and its fields.  Or, if you can draw it in google slides, we can share it w/ Vasily and discuss it.

>   • The only places where we have accessed bi_io_vec directly are merge_data and copy_pages.

OK good.  So that narrows things down.

> These functions take one new page and bio as input and copy the first page from bio to the new page.
1. It’s wrong to only look at the first page, right?  The vector may have multiple pages, yes?
2. Do we copy the entire page we get, or the start/end part out of it.  If we copy the whole page and ignore the start/end pointers, that’s wrong.

> This function has to be called several times from inside the loop to copy each bio_vec array.

Vasily mentioned that we’ll never get more than 4kb from above.  But, is it possible that we will get 4kb that’s NOT aligned on page boundary?  If so, we will have to turn that into two separate lower BIOs.  Vasily, do you know if that’s possible?

Also, even if we only get no more than 4kb from the upper layer, and let’s say it’s split over several smaller data units: is it possible that those will NOT be contiguous ranges of data?  If so, that’ll make it difficult for us to do the read-on-write (it’ll be fragmented).

> Or would it be wiser to call the function once and iterate from inside the function?
> Thoughts?

Venkat, I think you need to write a “validate bio” function that we will call before we clone or access the bio we get.  That function should check for all sorts of conditions that our code does not handle, or only verify what we DO handle, and otherwise BUG_ON() any bad condition.  That way, we write the code with certain assumptions, and if any of those assumptions are violated, we oops with a good printk message — giving us a chance to enhance our code to deal with any strange condition (fragmented bio, misaligned one, etc.).

Erez.

**Comment by venkrishraja** [comment]
1. Every bio does require a vector array and we need to create a vector of just one element at least, stuff that page into it and take care of other parameters.
2. Yes it’s wrong to only look at the first page.  The vector may have multiple pages
3. No, we do not copy the whole page. We only copy from the offset and till the length.
4. Each segment of data inside each bio_vec structure is contiguous but multiple bio_vecs are non-contiguous with one another.

**Comment by sectorsize512** [comment]
Just a clarification in accordance with my understanding. We always get a bio which is of maximum of 3.5KB or less (if it is 4KB then read-on-write path should not be called). The bio's vector can potentially contain 1, 2, ..., or 7 segments in this case. E.g., let's consider two imaginable scenarios of bio of len 3KB:

1) bio with 1 segment that points to a single page, offset within the page is (let's say) 0 and len is 3KB.

2) bio with 3 segments, each pointing to separate pages with offset of (let's say) 2KB and len of 1KB. (Notice, 3 \* 1KB = 3KB).

Now, the _clone_ bio that we create should always be of 4KB size. How many segments to have - it is absolutely up to us - but I would argue that having just 1 segment that points to one full page is more logical.

HTH,
Vasily

**Comment by ezk** [comment]
Yes, of course.  My concern is whether the three segments in your example are fragmented, with logical (LBA) “gaps” in between.  So I asked Venkat today to do some testing and printk’s, even oops if he finds something we don’t yet know how to handle.

Erez.

**Comment by sectorsize512** [comment]
Aah, I see. There cannot be LBA gaps. Every bio points only to one LBA offset and one length. Segments are only for gathering multiple memory regions.

**Comment by venkrishraja** [comment]
I have been doing some testing.

I don't think we need to assert whether size is less than 4KB because the read on write path is called only when the size is less than 4KB. I've checked as well. It is lesser than 4KB each time.

I also tested the page index as suggested to check for any logical gaps. As mentioned by Vasily there are no LBA gaps and the indexes are continuous.

So I am thinking of editing the merge_data() and copy_pages() functions by introducing a 'bio_for_each_segment' macro to loop through the bio_vec arrays.

One weird thing I noticed is that the bio_vcnt (which is the no.of bio_vec elements in the bi_io_vec array) of the bio has the value 0. Shouldn't it be, at the very least, 1?

**Comment by ezk** [comment]
> On Apr 27, 2016, at 5:05 AM, Venkatakrishnan notifications@github.com wrote:
> 
> So I am thinking of editing the merge_data() and copy_pages() functions by introducing a 'bio_for_each_segment' macro to loop through the bio_vec arrays.

Don’t “introduce” a new macro: USE the existing macros.

> One weird thing I noticed is that the bio_vcnt (which is the no.of bio_vec elements in the bi_io_vec array) of the bio has the value 0. Shouldn't it be, at the very least, 1?

Is it in our clone bio or the one we get from above?  If the one above, that would be odd; if it’s ours, then maybe we don’t init it correctly.

Also check how the various macros iterate over the bio_vec elements: exactly where they get the counter?

Erez.

**Comment by venkrishraja** [comment]
Yes I meant to use the existing macro.

Our clone bio has proper bi_vcnt. I am talking about the one we get above.

**Comment by ezk** [comment]
Strange, that our bio has a proper bi_vcnt but the one they give us has it set to zero?!  Is it always zero or some of the time.

Erez.

**Comment by venkrishraja** [comment]
All the loops use the bi_size i.e the residual IO to be processed as the condition for termination and loop till it becomes 0.

The bi_vcnt is always 0. Also, our clone bio has proper bi_vcnt because we are calling bio_add_page() which increments the bi_vcnt. I am not able to find any other place where the bi_vcnt is getting modified.

I also found this comment in block/bio.c

> Lastly, bi_vcnt should not be looked at or relied upon by code
> that does not own the bio - reason being drivers don't use it for
> iterating over the biovec anymore, so expecting it to be kept up
> to date (i.e. for clones that share the parent biovec) is just
> asking for trouble and would force extra work on
> __bio_clone_fast() anyways.

**Comment by ezk** [comment]
OK so we can’t rely on bi_vcnt to iterate over the vector, but rather on bi_size.  All the more reason to use THEIR macros and not try to DIY.

Erez.

**Comment by venkrishraja** [comment]
I made a few changes to merge_data() and copy_pages() as we discussed on Monday and tried running mkfs.xfs.

Good news is that there is no crash anymore :)
But it is taking a lot of time(5-6 seconds) to complete.
I'll be submitting my proper patch in some time.

I have attached a diff file for your reference.
[diff.txt](https://github.com/dmdedup/dmdedup/files/241868/diff.txt)

**Comment by sectorsize512** [comment]
That's great! Venkat, for future, when you generate a patch (diff.txt in
this case), please, use diff's "-up" parameters. This will generate a patch
in conventional format for reviewing.

Vasily

**Comment by ezk** [comment]
Nice.  Please submit a properly formatted patch for review asap.

Next, we’ll need to do more thorough testing.  Vasily, what do you recommend?  Use filebench inside dmdedup?  Use xfstests?

After that, I want to do a lot of code cleanup and documentation.  We’ll generate a large series of bisectable small patches, and carefully test each one to ensure we didn’t break anything.

Then, we’ll need to port the code to the latest kernel.

Erez.

**Comment by venkrishraja** [comment]
Sure.
I have submitted a new pull request with proper patch.
Kindly update me on whether to run filebench/xfstests and how to get started.

Venkat.

**Comment by ezk** [comment]
To start with, speak with Ming and Zhen on how to run FB and xfstests.  You’ll run them in a VM for now.

Erez.

**Comment by sectorsize512** [comment]
Generally, any file system or block level benchmark is a good start. Here
is one more to pick from:

https://github.com/linux-test-project/ltp/tree/master/testcases/kernel/fs

I think you can start with a benchmark of your choice and then gradually
extend the list of benchmarks your testing scripts support. To cover many
codepaths, it is desirable for a benchmark to support:

1) integrity checking
2) duplicate generation

Vasily

**Comment by ezk** [comment]
I used to use LTP for years, but moved to xfstests: most kernel developers have moved too.  It seems the xfstesst is better supported and more acceptable these days…

Erez.

**Comment by venkrishraja** [comment]
Sure, I'll get started with the benchmarks.

I have submitted the proper patch via a new pull request.
Is there anything else that you suggest for checking the fix?
Like creating/modifying files in the newly formatted xfs device?

**Comment by ezk** [comment]
Yes, start w/ some simple tests: mount the f/s, try to write/create/delete some files.  Maybe just download the kernel source and try to compile it, just to see a few basic f/s ops.  And try fsck too to see that it passes w/ corruptions.

Erez.

**Comment by venkrishraja** [comment]
I have submitted a new pull request with modified code.
I tried formatting ext4 with the patch. There is no crash and I am able to mount.

I tried fsck and mounting the f/s.
I get the following error while mounting.

> mount /dev/mapper/mydedup on /xfs failed: Structure needs cleaning

I have attached the logs for fsck and mounting.
[fsck.txt](https://github.com/dmdedup/dmdedup/files/248300/fsck.txt)
[mount.txt](https://github.com/dmdedup/dmdedup/files/248301/mount.txt)

**Comment by ezk** [comment]
What is /xfs?  Is that just a pathname?  Is that used for XFS or ext4?  if it’s just a mount point, please call it something else that’s more self-explanatory.

After a successful mkfs.ext4, did you try to fsck the ext4 partition?  Did that give you a corruption error?  If so, then we definitely now have a data corruption.

We discussed it earlier this week: you were going to check and see if we’re copying the less-than 4KB data into the 4KB read-on-write page, at the right starting offset; and doing so contiguously?  You should add maybe some printfs to see where we’re starting to read/write data into the 4KB page we read from below.

Also verify that in the full 4KB page, we’re writing data correctly at the starting offset.  I don’t know if in the code path that deals with exactly 4KB data — does that data come to us already aligned on 4KB boundaries, or not?  You can add a BUG_ON to see if that’s true or not.

Erez.

**Comment by venkrishraja** [comment]
It's just a mount point.
fsck for ext4 doesn't give a corruption error. It says "clean". I am able to mount as well.

**Comment by ezk** [comment]
Ok so let’s find out what’s going on w/ xfs: the corruption has to do with partial reads/writes.

Erez.

**Comment by venkrishraja** [comment]
We are calculating the offset as

> sector_t bi_sector = bio->bi_iter.bi_sector;
> position = sector_div(bi_sector, dc->sectors_per_block);
> dest_page_vaddr = page_address(page) + to_bytes(position);

So far my printk's suggest that the destination address is correct.
Might the corruption be because we are not replicating all the flags/fields from bio to clone?

**Comment by ezk** [comment]
Well, there’s some sort of data corruption, and we need to find out what’s going on.  You need to do some tests.  Start w/ the C program Vasily produced before — it was helpful.  Then do this:
1. get a dmdedup device ready.  Don’t format it.
2. zero out the device entirely:
   dd if=/dev/zero bs=4k of=/dmdedup/dev/path
3. write a C program that p/writes small but well known data items directly into the device.  For example, write the number X into offset X of the device.  Keep doing that randomly.  Then, read back the device sequentially, and see everywhere if you have the right value X in offset X — if it’s not, it’ll give you an idea what kind of corruption we have, and it’ll confirm we have a corruption.

Here’s some pseudo code
- find the size of your device in bytes — say X
- divide that size by sizeof(int) — say Y
- pick a random number b/t 0 and Y — call it R
- lseek(R \* sizeof(int)) — seek to aligned-word offset
- write(fd_of_dev, (void *) &R, sizeof(int))
- repeat last 3 steps many times — you can try sequentially and randomly

after you do the above enough time, stop and re-read the device one int (4 bytes) at a time, and see if in offset N of the device, you find the number N.

Erez.

**Comment by venkrishraja** [comment]
I followed all the steps. Whatever value I set to R(set the value even manually instead of random), the following code doesn't read the R value that was written, it just reads 0 into "buf".

lseek(fd, R \* sizeof(int), SEEK_SET);
write(fd, &R, sizeof(int));
lseek(fd, R \* sizeof(int), SEEK_SET);
read(fd, &buf, sizeof(int));
printf("%d ",buf);

**Comment by ezk** [comment]
Do a printf inside the kernel code, right before write data to the bio: are we writing all zeros?  Are writing using the correct offset?  Have we properly memcpy’d the data from the upper bio to ours?

Erez.

**Comment by ezk** [comment]
What is “fd” below: is it a single file or did you open the full /dev?

I’m curious to do this experiment: (1) write some data, say all 1s to some small file.

(2) run something like “dd if=/dev/our-dmdedup-dev bs=4k” and see if you see any pattern of 1s anywhere.  I want to know if we never write any data, or we’re writing it in the wrong offset?

Erez.

**Comment by venkrishraja** [comment]
fd is the full dev.

Where do I write this file to? 
Since mounting the dmdedup device is giving an error, I cannot create a new file without mounting.
Is there any other way to try this?

**Comment by ezk** [comment]
Open the raw device directly and write to it.  You don’t even need to mkfs it first.  Yan start by zeroing out the dev: dd if=/dev/zero of=/dev/dmdedup bs=4k and then read it back.  Then you can open() /dev/dmdedup, lseek, write data, and try to read it back.

Erez.

**Comment by venkrishraja** [comment]
I did a git checkout and started with the basic code.
I applied my patches one after another. 

Now there is no metadata corruption and I am able to mount, create, copy, delete files.
I repeated this a few times and also checked "xfs_check" and "xfs_repair -n" commands.
No errors/corruption detected.

**Comment by ezk** [comment]
Good.  Can you find out what’s changed: was there any extra code in your “bad” git repo that was causing this problem?

Can you please [re]submit the (clean) patches for review?

Erez.

**Comment by venkrishraja** [comment]
Yes, I did not replace 'my_zero_fill_bio' with 'zero_fill_bio' which seems to do a lot more.
A lot of functions were 'noinline' before as well.

But I am not sure. It might be something else that I changed.
But I have checked repeatedly to ensure there is no corruption.

**Comment by ezk** [comment]
Ok if that works, it works.  Please submit the patches for review asap.

Vasily, can you review the patches and let us know if you have any concerns?  I’d like to have those patches in github soon, as I’d like to play w/ dmdedup myself over the summer.

Thanks,
Erez.

**Comment by venkrishraja** [comment]
I have submitted the latest patch for review.
Please let me know if something needs to be changed.

**Comment by ezk** [comment]
Why do you still use my_fill_zero?!  This seems to have caused us trouble, so let’s nuke it and use the standard fill_zero routine from the bio code.

Erez.

**Comment by venkrishraja** [comment]
I have resubmitted the fix with proper Git header.
I'll submit the zero_fill_bio fix after this fix is merged.

**Comment by ezk** [comment]
No, don’t wait.  Submit the zero_fill patch now.  No need to wait.  There can be N patches being reviewed at any time.

Erez.

**Comment by venkrishraja** [comment]
No if I try to submit this patch as a new pull request, GitHub merges this commit with the previous unmerged commit together. That's why I want to wait till the previous one's merged.

**Comment by ezk** [comment]
Really?  That seems odd and very limiting.  Most systems I know allow one to submit a series of patches, called a “patchset”, and review it as a unit.

Vasily, is there something we need to configure differently on github?

Erez.

**Comment by sectorsize512** [comment]
The thing is that merge requests are for branches, not for the commits. So,
if there are two commits on the branch, there still will be a single pull
request (but for a branch with two commits). In order to have multiple
independent merge requests one has to put those commits in different
branches. Venkat, you can do that in the future (or now, for the sake of
the practice), but I'm going to approve the pending merge request in a
second.

Vasily

**Comment by sectorsize512** [comment]
Just wanted to let you know that the merge request contained both commits:
1) bio_iter and 2) zero_fill. So,  by approving this merge request I made
both commits available in the main repository.

Vasily

**Comment by venkrishraja** [comment]
Thank you, Vasily.
Can we close this thread and open new ones for future issues?

**Comment by sectorsize512** [comment]
Fixed with 7f9e497
